### DISCUSSION TITLE
Jetpack Compose / AsyncImage: Specify placeholder size when loaded from memory

### DISCUSSION THREAD

**Discussion by svenjacobs** [coil-kt/coil#1152] [Q&A] [answered]
I have a classic list / detail relationship: List of items are displayed, when clicked a detail screen is opened. List items and detail both contain images. In order to improve the user experience, I want to use the smaller image from the list view as the placeholder image of the larger image in the detail view, until the high resolution image is loaded.

For the list items I use `memoryCacheKey("SOME_UUID")` and for the detail screen I use `placeholderMemoryCacheKey("SOME_UUID")`. This works however I noticed the following: The placeholder image does not have the size of the detail image. It actually has the size of the list view. In other words: If list image is `50x50dp` and detail is `100x100dp`, a `50x50` placeholder is loaded into the `100x100` image. Then, when the high resolution image finished loading, the placeholder is replaced by the bigger image which causes a "leap" of the UI. Is it possible to have `AsyncImage` scale the placeholder image to the size of the container? I would rather have a blurred, low-resolution placeholder in the correct size than a leap.

Coil version: `2.0.0-alpha08`

**Comment by svenjacobs** [comment]
Also I noticed despite `crossfade(true)` there is no crossfade between placeholder and actual image.

**Comment by colinrtwhite** [comment, marked as answer]
@svenjacobs `ContentScale.FillBounds` scales the image to fill the bounds.

**Reply by svenjacobs** [reply]
@colinrtwhite Thanks for your answer. Unfortunately `FillBounds` does not work as it does not respect the ratio and images are distorted. What I actually want to achieve is "just" scaling the placeholder image to the size of the actual image while taking `contentScale` into account. Is this possible?

**Reply by svenjacobs** [reply]
Reading my initial post again I understand that the example sizes that I mentioned might be misleading. The images do not necessarily have square sizes. They can have any ratio. Therefore `FillBounds` doesn't work here.

**Reply by colinrtwhite** [reply]
`ContentScale.Fit` will preserve aspect ratio. It's not possible to scale the placeholder separately from the success/error image using `AsyncImage`; you might be able to do that with `rememberAsyncImagePainter`.

**Reply by svenjacobs** [reply]
Just for completeness, this is the solution that I went for:

```kotlin
/**
 * Similar to [AsyncImage] but displays a placeholder image on top of actual image during load.
 *
 * The difference to specifying [ImageRequest.Builder.placeholder] or [ImageRequest.Builder.placeholderMemoryCacheKey]
 * for a regular `AsyncImage` is that the placeholder image will be scaled to the same size as the
 * actual image. If the placeholder bitmap is smaller than the actual image, a regular `AsyncImage`
 * would not scale the placeholder and the user notices a disruptive change of image size.
 *
 * If `placeholderImageCacheKey` is null or an image with that key is not found in the memory cache,
 * a loading indicator is displayed instead.
 */
@Composable
fun LoadableAsyncImage(
    modifier: Modifier = Modifier,
    model: Any?,
    contentDescription: String?,
    placeholderMemoryCacheKey: String? = null,
    loadingIndicatorSize: Dp = 40.dp,
    contentScale: ContentScale = ContentScale.Fit,
) {
    val imageLoader = LocalContext.current.imageLoader
    var placeholderBitmap by remember(placeholderMemoryCacheKey) { mutableStateOf<Bitmap?>(null) }
    var isLoading by rememberSaveable(model) { mutableStateOf(true) }

    LaunchedEffect(placeholderMemoryCacheKey) {
        placeholderMemoryCacheKey?.let {
            placeholderBitmap =
                imageLoader.memoryCache?.get(MemoryCache.Key(placeholderMemoryCacheKey))?.bitmap
        }
    }

    Box(
        modifier = modifier,
        contentAlignment = Alignment.Center,
    ) {
        AsyncImage(
            modifier = Modifier.fillMaxSize(),
            model = model,
            contentDescription = contentDescription,
            contentScale = contentScale,
            onSuccess = { isLoading = false },
        )

        AnimatedVisibility(
            visible = isLoading,
            enter = fadeIn(),
            exit = fadeOut(),
        ) {
            if (placeholderBitmap == null) {
                LoadingIndicator(
                    size = loadingIndicatorSize,
                )
            } else {
                Image(
                    modifier = Modifier.fillMaxSize(),
                    bitmap = (placeholderBitmap as Bitmap).asImageBitmap(),
                    contentDescription = contentDescription,
                    contentScale = contentScale,
                )
            }
        }
    }
}
```

**Reply by isaialvarado** [reply]
> `ContentScale.Fit` will preserve aspect ratio. It's not possible to scale the placeholder separately from the success/error image using `AsyncImage`; you might be able to do that with `rememberAsyncImagePainter`.

At least in 2.2.2, if you use `rememberAsyncImagePainter` for a placeholder, the ContentScale for that placeholder `rememberAsyncImagePainter` is ignored.

**Comment by isaialvarado** [comment]
I was already using a custom `ContentScale` with state, so a workaround for me was to force the scale to be based on a `Size` I am in control of. Example of a custom `ContentScale`

```
class MyContentScale(
    private val theRealDestinationSize: State<Size>,
    private val contentScale: ContentScale
) : ContentScale {
    override fun computeScaleFactor(srcSize: Size, dstSize: Size): ScaleFactor {
//        ignore dstSize, instead use the REAL destination size
        return contentScale.computeScaleFactor(srcSize, theRealDestinationSize.value)
    }
}
```

I use a similar `ContentScale` on my `AsyncImage` while also using the placeholder parameter on `AsyncImage`. This makes sure my placeholder and image are loaded at the same scale, even if they are different resolutions.
